ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll

```diff
@@ -230,31 +334,128 @@ class Boolean extends boolean {
 
 private import SummaryComponentStack
 
+pragma[nomagic]
 private predicate hasStoreSummary(
-  SummarizedCallable callable, DataFlow::ContentSet contents, SummaryComponent input,
-  SummaryComponent output
+  SummarizedCallable callable, DataFlow::ContentSet contents, SummaryComponentStack input,
+  SummaryComponentStack output
 ) {
-  callable
-      .propagatesFlow(singleton(input),
-        push(SummaryComponent::content(contents), singleton(output)), true)
+  not TypeTrackingStep::suppressSummary(callable) and
+  callable.propagatesFlow(input, push(SummaryComponent::content(contents), output), true) and
+  not isNonLocal(input.head()) and
+  not isNonLocal(output.head())
 }
 
+pragma[nomagic]
 private predicate hasLoadSummary(
-  SummarizedCallable callable, DataFlow::ContentSet contents, SummaryComponent input,
-  SummaryComponent output
+  SummarizedCallable callable, DataFlow::ContentSet contents, SummaryComponentStack input,
+  SummaryComponentStack output
 ) {
+  not TypeTrackingStep::suppressSummary(callable) and
+  callable.propagatesFlow(push(SummaryComponent::content(contents), input), output, true) and
+  not isNonLocal(input.head()) and
+  not isNonLocal(output.head())
+}
+
+pragma[nomagic]
+private predicate hasLoadStoreSummary(
+  SummarizedCallable callable, DataFlow::ContentSet loadContents,
+  DataFlow::ContentSet storeContents, SummaryComponentStack input, SummaryComponentStack output
+) {
+  not TypeTrackingStep::suppressSummary(callable) and
   callable
-      .propagatesFlow(push(SummaryComponent::content(contents), singleton(input)),
-        singleton(output), true)
+      .propagatesFlow(push(SummaryComponent::content(loadContents), input),
+        push(SummaryComponent::content(storeContents), output), true) and
+  not isNonLocal(input.head()) and
+  not isNonLocal(output.head())
+}
+
+/**
+ * Gets a content filter to use for a `WithoutContent[content]` step, or has no result if
+ * the step should be treated as ordinary flow.
+ *
+ * `WithoutContent` is often used to perform strong updates on individual collection elements, but for
+ * type-tracking this is rarely beneficial and quite expensive. However, `WithoutContent` can be quite useful
+ * for restricting the type of an object, and in these cases we translate it to a filter.
+ */
+private ContentFilter getFilterFromWithoutContentStep(DataFlow::ContentSet content) {
+  (
+    content.isAnyElement()
+    or
+    content.isElementLowerBoundOrUnknown(_)
+    or
+    content.isSingleton(any(DataFlow::Content::UnknownElementContent c))
+  ) and
+  result = MkElementFilter()
+}
+
+pragma[nomagic]
+private predicate hasWithoutContentSummary(
```

Comment:
Does this work with compound `WithoutContent`s (see https://github.com/github/codeql/pull/10691)?

Reply:
If at most one of those `WithoutContent` has an associated `ContentFilter` then yes, because the rest are ignored and treated as ordinary flow (in `evaluateSummaryComponentStackLocal`).